lib/fathead/python/redirect.py

```diff
@@ -0,0 +1,170 @@
+        Args:
+            input_obj: TSV string or list of data.
+
+        Returns:
+            List of data
+        Raises:
+            Throws BadEntryException if data is invalid.
+        """
+        if isinstance(input_obj, str):
+            processed = input_obj.split('\t')
+            self.data = processed
+        elif isinstance(input_obj, list):
+            self.data = input_obj
+
+        try:
+            self.key = self.data[0].strip()
+            self.entry_type = self.data[1].strip()
+            self.reference = self.data[2].strip()
+            if len(self.data) > 3:
+                self.abstract = self.data[11].strip()
+                self.anchor = self.data[12].strip()
+            elif self.entry_type == 13 and len(self.data) != 13:
+                raise BadEntryException
+        except Exception as e:
+            raise BadEntryException('Article had invalid number of elements.')
+
+        if self.entry_type == 'A':
+            self.parse_alternative_keys()
+
+        return self.data
+
+    def parse_alternative_keys(self):
+        """
+        Find alternative keys to use in generated redirects
+
+        Returns:
+            Set of possible redirect entries
+        """
+        self.alternative_keys = set()
+
+        if '.' in self.key and self.entry_type == 'A':
+            key_arr = self.key.split('.')
+            method_name = key_arr[-1]
+            package_name = key_arr[0]
+            new_key = "{} {}".format(package_name, method_name)
+            self.alternative_keys.add(new_key)
+            self.alternative_keys.add(method_name)
+        return self.alternative_keys
+
+    def get_data(self):
+        return self.data
+
+    def get_key(self):
+        return self.key
+
+    def get_type(self):
+        return self.entry_type
+
+    def get_redirects(self):
+        redirs = []
+
+        for alt_key in list(self.alternative_keys):
+            entry = Entry([
+                alt_key,
+                'R',
+                self.key
+            ])
+            redirs.append(entry)
+
+        return redirs
+
+    def get_alternatives(self):
+        return self.alternative_keys
+
+    def get_entry(self):
+        return '\t'.join([
+            self.key,             # title / key
+            self.entry_type,      # entry type
+            self.reference,       # no redirect data
+            '',                   # ignore
+            '',                   # no categories
+            '',                   # ignore
+            '',                   # no related topics
+            '',                   # ignore
+            '',                   # add an external link back to page
+            '',                   # no disambiguation
+            '',                   # images
+            self.abstract,        # abstract
+            self.anchor           # anchor to specific section
+        ])
+
+    def __str__(self):
+        return self.get_entry()
+
+def generate_redirects(f):
```

Comment:
I just want to make sure I understand this function first. 
1.  go through all the lines in output.txt
2. create an entry object for each. This contains all the redirects
3. add the key to `output` if it doesn't exist already
4. go through each redirect and add those to `output` if they don't exist already

Does that sound right?